optuna/pruners/hyperband.py

```diff
@@ -0,0 +1,156 @@
+            :math:`\\eta` in the paper. See the details for
+            :class:`~optuna.pruners.SuccessiveHalvingPruner`.
+        min_early_stopping_rate_low:
+            The start point of the minimum early stopping rate for ``SuccessiveHalvingPruner``.
+        min_early_stopping_rate_high:
+            The end point of the minimum early stopping rate for ``SuccessiveHalvingPruner``.
+    """
+
+    def __init__(
+            self,
+            min_resource=1,
+            reduction_factor=3,
+            min_early_stopping_rate_low=0,
+            min_early_stopping_rate_high=4
+    ):
+        # type: (int, int, int, int) -> None
+        self._pruners = []  # type: List[SuccessiveHalvingPruner]
+        self._reduction_factor = reduction_factor
+        self._resource_budget = 0
+        n_pruners = min_early_stopping_rate_high - min_early_stopping_rate_low + 1
+        self._n_pruners = n_pruners
+        self._bracket_resource_budgets = []  # type: List[int]
+
+        _logger.debug('Hyperband has {} brackets'.format(self.n_pruners))
+
+        for i in range(n_pruners):
+            bracket_resource_budget = self._calc_bracket_resource_budget(i, n_pruners)
+            self._resource_budget += bracket_resource_budget
+            self._bracket_resource_budgets.append(bracket_resource_budget)
+
+            # N.B. (crcrpar): `min_early_stopping_rate` has the information of `bracket_index`.
+            min_early_stopping_rate = min_early_stopping_rate_low + i
+
+            _logger.debug(
+                '{}th bracket has minimum early stopping rate of {}'.format(
+                    i, min_early_stopping_rate))
+
+            pruner = SuccessiveHalvingPruner(
+                min_resource=min_resource,
+                reduction_factor=reduction_factor,
+                min_early_stopping_rate=min_early_stopping_rate,
+            )
+            self._pruners.append(pruner)
+
+    def prune(self, study, trial):
+        # type: (Study, FrozenTrial) -> bool
+
+        i = self.get_bracket_id(study.study_name, trial.number)
+        _logger.debug('{}th bracket is selected'.format(i))
+        bracket_study = _BracketStudy(study, i)
+        return self._pruners[i].prune(bracket_study, trial)
+
+    def _calc_bracket_resource_budget(self, pruner_index, n_pruners):
+        # type: (int, int) -> int
+        n = self._reduction_factor ** (n_pruners - 1)
+        budget = n
+        for i in range(pruner_index, n_pruners - 1):
+            budget += n / 2
+        return budget
+
+    def get_bracket_id(self, study_name, trial_number):
+        # type: (str, int) -> int
+        """Computes the index of bracket for a trial of ``trial_number``.
+
+        The index of a bracket is noted as :math:`s` in
+        `Hyperband paper <http://www.jmlr.org/papers/volume18/16-558/16-558.pdf>`_.
+        """
+
+        n = hash('{}_{}'.format(study_name, trial_number)) % self._resource_budget
```

Comment:
A question rather than a suggestion but would it simplify passing a `(study_name, trial_number)` tuple?

Reply:
Do you mean the following?
```suggestion
        n = hash((study_name, trial_number)) % self._resource_budget
```

Makes sense.

Reply:
It looks better to me.

Reply:
This is just a question. May I have the reason why it calculates a hash value with `study_name` ? It looks essentially the same with `n = trial_number % self._resource_budget` except for including its randomness.

I guess this change is based on following lines by @sile:
https://github.com/optuna/optuna/pull/301/files#diff-45eb5bd2359f1b7f13b35b43caecb92bR55-R59

Reply:
@c-bata 

"including its randomness" is the reason. However, perhaps there is no strong rationale to introduce the randomness, so I think that we could reconsider adopting more simple `n = trial_number % self._resource_budget` as the bracket selection logic.

Reply:
Okay. I also don't have a strong opinion about this. I just tend to choice simpler implementation. Thank you for your answer!

(I'll choice `n = trial_number % self._resource_budget` at https://github.com/c-bata/goptuna/pull/63)

Reply:
Hello from 2021!
I think there could be an error about using hash for computing the bracket.
To be specific, I conducted the parallel Optuna experiments using the Hyperband Pruner.
Then, the hash gives different value for the same input when executed on different terminal (different python script). For example, a trial _n_ could be assigned to the bracket 1 at process 1 and the bracket 2 at process 2 at the same time, which ruins the benefits of Hyperband over the SuccesiveHalving.
As you can see [here](https://docs.python.org/3.3/using/cmdline.html), it was due to the update that hash randomization is turned on after python version 3.3.
Of course, the randomization could be turned off by setting the _PYTHONHASHSEED=0_ when executing the python script. Since I did not see any warning or suggestion about this issue at the documentation, I would like to ask if my assertion was legit.
I include some screenshots that the trials belonging to the same bracket at each parallel process were different.

 ![Process 1 - trials for bracket 1](https://user-images.githubusercontent.com/55653250/140302703-a7ea8e8a-0f9d-4be1-ae1c-246686535c17.png)

![Process 2 - trials for bracket 1](https://user-images.githubusercontent.com/55653250/140302720-94e346e1-0984-4278-a278-005534efe9cc.png)

Reply:
@donny8 would it maybe be good to write this "problem" as a dedicated issue?

Reply:
@PhilipMay Oh I was not sure if I did understand the circumstance correctly. I would make a new issue about this problem. Thanks for the suggestion.